### ISSUE TITLE
Transition removed from rule

### ISSUE THREAD

**Issue by hummal** [pocketjoso/penthouse#164]
Hey!

At first: Thanks for your effort!

I found a (maybe) bug. Iam not sure if it is penthouse related or phantomjs.

I got this rule:
`.swt-Map_ChildlinkPinIcon {
    display: inline-block;
    vertical-align: middle;
    text-align: center;
    opacity: 0;
    width: 0;
    -webkit-transition: opacity 0.25s ease-in-out, margin 0.25s ease-in-out;
    -moz-transition: opacity 0.25s ease-in-out, margin 0.25s ease-in-out;
    -o-transition: opacity 0.25s ease-in-out, margin 0.25s ease-in-out;
    transition: opacity 0.25s ease-in-out, margin 0.25s ease-in-out;
}`

after Penthouse processing it, the only thing left is:
`.swt-Map_ChildlinkPinIcon {
    display: inline-block;
    vertical-align: middle;
    text-align: center;
    opacity: 0;
    width: 0;
}`

As you can see, the transitions are gone. I really dont know why.

**Comment by pocketjoso** [comment]
Hi @hummal,
It's not a bug, it is (truly) an undocumented feature. It was introduced [https://github.com/pocketjoso/penthouse/pull/100].

Since transitions are not part of the first render, we can get to the first render even faster further by removing them from the critical css, and hence reducing the size.

Does this cause any issues on your end? The transitions will come back with your full css loading.

**Comment by hummal** [comment]
Hey Jonas,
thank you for fast response and the explanation. 
Indeed I have some issues because of that. To explain what Iam trying to do. 
I have a lot different merged css files. Every one of them is for a specific amount of pages (40k+) which belong to a category. I need to minimize the critical css of all of these pages and build up an optimized file for every css file given. At the end I will get my example.css and a example.atf.css and example.rest.css. All the rules that are critical will be in the atf.css. Also these rules are removed of the original css and result in the rest.css. 
The problem is, that because of the removing the original rules, the transition is gone. 
Maybe I could also check if that rule content matches completely with the original rule and if not keep both, but this would result in a larger rest.css and is not what I need. 

The current status of my project is a 100/100 with pagespeed. I would like to keep that. The css is really big (1MB) and I need to keep the focus und size. 

I would love to have a option to turn off the transitions. What do you think?

**Comment by pocketjoso** [comment]
Sorry @hummal, I thought I had replied to your response, must have missed to do so.

In general you don't want to extract the critical css from any other files, as the critical css is inlined in the html and hence never gets cached - you want to instead be able to rely on the cached full css as soon as the user has loaded it.

If you left the original css untouched (or in your case just as the result of concatenating all the page instances of the same category), then you should have the transition there in the full page css. Would that be enough for your use case?

I realise quite some time has passed since you asked this, let me know if this is still relevant for you.

Cheers

**Comment by hummal** [comment]
Checked and everything is fine. Maybe It seems to be a false positive. 

Thanks for your effort!

**Comment by hummal** [comment]
After I was needed to enhance my code, I remembered what my problem was. 

I use penthouse to check my _full.css_ against  number of sites. I save all results ( also duplicates ) in one file called _atf.css_. Now I need to remove the duplicates.
The result ist my final _atf.css_ for inlining into the document head. Works fine 'til now.

But I also need the rest of my _full.css_, but I don't want it to have all the _atf.css_ rules doubled. So I check every single rule of both files and remove the matching rules from _full.css_. 

Problem: This wont work if I have a rule with transitions. Because the rule string has changed. And with rule I mean the complete rule ( .test .test2 { color: blue; transition: ... } )

Possible solution on my side: merge same rules to one and remove the whole rule just by the selector string. But I need to rely on different rules to be on different places in my code. I don't know if a specific rule have to be at the end and the same rule with different content needs to be on the beginning.
Thinking about this example:
```css
.test {
  color:blue;
  margin: 10px;
}
.anotherRule {}

.test2 {
  color: red;
}
```
I wouldn't know which color is right, so merging is not possible. 

My task is to keep the remaining css as tiny as possible. 

Thats why I would love to have the **option not to remove transitions and co**

**Comment by pocketjoso** [comment]
> I use penthouse to check my full.css against number of sites.

Do you mean pages? If you do mean sites, please explain.. why would you reuse the same styles for multiple sites?

> But I also need the rest of my full.css, but I don't want it to have all the atf.css rules doubled.

Like I said before:
> In general you don't want to extract the critical css from any other files, as the critical css is inlined in the html and hence never gets cached - you want to instead be able to rely on the cached full css as soon as the user has loaded it.

Removing critical css from the full css only makes sense if you serve the same critical css on every page, because only then would you still get exactly the same full css on every page, so it could be re-used. But combining critical css from multiple pages increases the size and makes the performance worse, so it's not something you want to do. At the point where you are combining critical css for a whole site you might almost as well just inline the whole full css instead - but you're not getting very good performance at this point.

With all this being said, I hope you can see why I don't really want to add support for this use case, because I think it risks misleading people into using critical css in a way they would be better of not doing.

**Comment by hummal** [comment]
You are right. I mean pages!

Iam working for a big company and we have 40k+ pages which are generated by database contents. I have 26 different generated css files for all of these pages. Depending on what is displayed. 
So I have one file (full.css) to extract my critical from. This is how penthouse works, right?
Additionally I need to check this 26 css files against a bunch of pages to get the most out of the css. Then, as I already said, I need to reduce the css to a minimum because there are duplicates in it. At the end the critical css is like I just ran penthouse once. So no need to talk about that.

But even tho I have the critical css now and it is good, I want to remove the critical css rules from my full.css to reduce its size.

If you dont want to add the option for me its ok, but I need to do a fork then. Because I can reach more with this option. There are about 3% code I can't remove. 

Maybe Iam not able to clearly tell you what I need. Sorry for that. Still good work. Thank you

**Comment by pocketjoso** [comment]
So let me see:

You're saying you're downloading different (_full_) css files for different pages, hence you never serve one combined css file for the whole site?

The next question is: for the groups of pages sharing the same stylesheet, for example hypothetical `stylesheet A` (one of the 26 you mentioned above) - will you serve them all the same critical css, or different?
- If you serving the same critical css.. depending on how many pages you're sharing it between, and how different they are - you will not get very good results as the critical css will grow very big.

- if - on the other hand - you serve _different_ critical css to the different pages in the same group: then what I've been saying before applies: you don't want to subtract the critical css, as it would create unique modified versions of the full stylesheet (`stylesheet A`) for each page, so no re-use, and no shared cache.

**Comment by hummal** [comment]
> will you serve them all the same critical css, or different?
Yes. I know that there will be more than the needed critical css for every page. But it is not possible for me to reduce this for every page. Because of the dynamic creation and user generated content. Thats why I have 1 of 26 files served for 1000+ pages each. These 1000+ pages will be part of a category. Like blog posts which have a common layout and so I only need to check against 1 blog page. (just for the context) Lets call that type of pages "Pagetype".

And no, the critical CSS won't grow that big in this case. I have up to 20% ununsed critical css. Which is not that much. Workflow:

1. critical css check stylesheet a against pagetype 1 > save to critical.css
2. check against stylesheet a against pagetype 2 > append to critical.css
3. go on for all pagetypes to check against stylesheet a
4. remove duplicate rules of critical.css
5. finished (max 20% overhead, which is ok for my use case) 100/100 still possible

I know you would understand much better if I would show what I've builded. And maybe someday I will be able to release it as open source but at the moment Iam not allowed to. :(

**Comment by pocketjoso** [comment]
> Because of the dynamic creation and user generated content. 

For dynamic content, normally you would use `forceInclude`, this is what is meant for. Do you know ahead of time what types of content users can inject? If so, just generate critical css for one page of your type, and foceInclude selectors for all dynamic content that could there in different variations.

Serving the same critical css for every page of a certain type is fine.
I would still strongly recommend agains trying to remove the critical css from your regular stylesheet.

**Comment by pocketjoso** [comment]
Closing due to inactivity - feel free to comment again and I can re-open if there are still things to discuss.

**Comment by hummal** [comment]
Sorry for being late. Vacation time!

I know about the feature to forceInclude. But Iam building a framework around penthouse to enable multi sites. So I had to build my own forceInclude. Next to forceExclude. 

Obviously Iam not able to explain what Iam doing. 
Why should I ever keep the extracted critical css in my main.css. It is ~50kb of already existing stylesheets. I will have doubled rules. In the actual case its right: transitions are missing. But I don't delete rules that doesnt match. 

That is the reason why I would love to have the option to keep transitions.

**Comment by pocketjoso** [comment]
> Why should I ever keep the extracted critical css in my main.css
I give up here - I don't think we're able to understand each other. I've tried explaining this too many times. :) Perhaps you have a legit use case, perhaps this is just miscommunication. Either way, I will not spend more time trying to explain why it is generally a bad idea trying to extract the critical css.

> That is the reason why I would love to have the option to keep transitions.

After #174 lands, feel free to make a PR. I suggest to add another Penthouse option to override [this list of patterns for properties to be removed](https://github.com/pocketjoso/penthouse/blob/chrome-headless/src/postformatting/index.js#L34-L40), so you could remove the transition removal, for your use case.

**Comment by hummal** [comment]
Check! Good idea with the list. Is more useful than just a option to keep transitions! Thanks

**Comment by pocketjoso** [comment]
#174 did land, and [penthouse 1.0 is out](https://medium.com/@pocketjoso/penthouse-1-0-official-release-ece995b0d29e) - so please upgrade before doing any further investigation.

**Comment by pocketjoso** [comment]
transition properties can now be kept, if necessary, via the `propertiesToRemove` property, added in #181.

**Comment by pocketjoso** [comment]
@hummal if you have other issues that remain, please open separate issues describing what they are. Thanks again for the PR!
